firestore/api/SolutionCounter/Program.cs

```diff
@@ -0,0 +1,137 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+using Google.Api.Gax;
+using Google.Cloud.Firestore;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace GoogleCloudSamples
+{
+    /// <summary>
+    /// To support more frequent counter updates, create a distributed counter.
+    /// Each counter is a document with a subcollection of "shards," and 
+    /// the value of the counter is the sum of the value of the shards.
+    /// </summary>
+    public class DistributedCounter
+    {
+        private static readonly Random s_rand = new Random();
+        private static readonly object s_randLock = new object();
+        private const string Usage = @"Usage:
+C:\> dotnet run command YOUR_PROJECT_ID
+
+Where command is one of
+    distributed-counter
+";
+        // [START fs_counter_classes]
+        /// <summary>
+        /// Shard is a document that contains the count.
+        /// </summary>
+        [FirestoreData]
+        public class Shard
+        {
+            [FirestoreProperty(name: "count")]
+            public int Count { get; set; }
+        }
+        // [END fs_counter_classes]
+
+        // [START fs_create_counter]
+        /// <summary>
+        /// Create a given number of shards as a
+        /// subcollection of specified document.
+        /// </summary>
+        /// <param name="docRef">The document reference <see cref="DocumentReference"/></param>
+        /// <returns>The <see cref="Task"/></returns>
+        private static async Task CreateCounterAsync(DocumentReference docRef, int numOfShards)
+        {
+            CollectionReference colRef = docRef.Collection("shards");
+            var tasks = new List<Task>();
+            // Initialize each shard with Count=0
+            for (var i = 0; i < numOfShards; i++)
+            {
+                tasks.Add(colRef.Document(i.ToString()).SetAsync(new Shard() { Count = 0 }));
+            }
+            await Task.WhenAll(tasks);
+        }
+        // [END fs_create_counter]
+
+        // [START fs_increment_counter]
+        /// <summary>
+        /// Increment a randomly picked shard by 1.
+        /// </summary>
+        /// <param name="docRef">The document reference <see cref="DocumentReference"/></param>
+        /// <returns>The <see cref="Task"/></returns>
+        private static async Task IncrementCounterAsync(DocumentReference docRef, int numOfShards)
+        {
+            int documentId;
+            lock (s_randLock)
+            {
+                documentId = s_rand.Next(numOfShards);
+            }
+            var shardRef = docRef.Collection("shards").Document(documentId.ToString());
+            await shardRef.UpdateAsync("count", FieldValue.Increment(1));
```

Comment:
This is just because I'm not that familiar with Firestore, but I'm a little worried still about how this handles concurrent modifications and whether the example makes sense or not then.
Imagine three different threads grab shard number 5 at the same time, that is when the counter is say, 3. They all increment the value by one, but I wonder, aftert the 3 increments, is the shard value 6 or 4? This depends on how UpdateAsync works but I'm not really that familiar with that. Maybe a test with only one shard and hundreds of simultaneous increments can shed some light on this.
My point is that, if in my example, the final result of the shard is 4, then the whole distributed counter example doesn't make a lot of sense as it stands.

Reply:
I run with hundreds of simultaneous increments and it's working fine.
I don't see any problem here because we increment counter by 1 with existing value.

Reply:
Perfect then.